pkg/kv/mem/store.go

```diff
@@ -146,72 +139,56 @@ func (s *Store) Delete(_ context.Context, partitionKey, key []byte) error {
 	}
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	combinedKey := combinedKey(partitionKey, key)
 
-	if _, found := s.m[combinedKey]; !found {
+	sKey := encodeKey(key)
+	if _, ok := s.m[string(partitionKey)][sKey]; !ok {
 		return nil
 	}
-	idx := sort.SearchStrings(s.keys, combinedKey)
-	if idx < len(s.keys) && s.keys[idx] == combinedKey {
-		s.keys = append(s.keys[:idx], s.keys[idx+1:]...)
-	}
-	delete(s.m, combinedKey)
+	delete(s.m[string(partitionKey)], sKey)
 	return nil
 }
 
 func (s *Store) Scan(_ context.Context, partitionKey, start []byte) (kv.EntriesIterator, error) {
 	if len(partitionKey) == 0 {
 		return nil, kv.ErrMissingPartitionKey
 	}
-	combinedKey := combinedKey(partitionKey, start)
 
 	return &EntriesIterator{
 		store:     s,
-		start:     combinedKey,
-		partition: partitionKey,
+		start:     start,
+		partition: string(partitionKey),
 	}, nil
 }
 
 func (s *Store) Close() {}
 
 func (e *EntriesIterator) Next() bool {
-	if e.err != nil {
-		return false
-	}
-	if e.start == "" {
-		e.entry = nil
+	if e.err != nil || e.start == nil { // start is nil only if last iteration we reached end of keys
 		return false
 	}
+
 	e.store.mu.RLock()
 	defer e.store.mu.RUnlock()
-	idx := sort.SearchStrings(e.store.keys, e.start)
-	if idx == len(e.store.keys) {
-		e.start = ""
-		e.entry = nil
-		return false
-	}
-	// point to the entry we found
-	combinedKey := e.store.keys[idx]
-	value := e.store.m[combinedKey]
 
-	partition := partitionKeyFromCombinedKey(combinedKey)
-	if !bytes.Equal(partition, e.partition) {
-		e.start = ""
-		e.entry = nil
-		return false
-	}
-
-	e.entry = &kv.Entry{
-		PartitionKey: partition,
-		Key:          keyFromCombinedKey(combinedKey),
-		Value:        value,
+	l := make([]*kv.Entry, 0)
+	if _, ok := e.store.m[e.partition]; ok {
+		for _, v := range e.store.m[e.partition] {
+			if bytes.Compare(v.Key, e.start) >= 0 {
+				entry := v
+				l = append(l, &entry)
+			}
+		}
 	}
-	// set start to the next item - nil as end indicator
-	if idx+1 < len(e.store.keys) {
-		e.start = e.store.keys[idx+1]
+	sort.Slice(l, func(i, j int) bool { return bytes.Compare(l[i].Key, l[j].Key) < 0 })
+	if len(l) <= EmptyResult+1 { // only 1 key >= start, set start to nil, so to indicate this is the last iteration
+		e.start = nil
+		if len(l) == EmptyResult { // No results
+			return false
+		}
```

Comment:
I think this is more readable - per your consideration
```suggestion 
	if len(l) == 0 {
		e.start = nil
		return false
	} 
	if len(l) == 1 {
		e.start = nil
		e.entry = l[0]
		return true
	}
	sort.Slice(l, func(i, j int) bool { return bytes.Compare(l[i].Key, l[j].Key) < 0 })
	e.start = l[1]
	e.entry = l[0]
	return true
```